[Python 3.9]  Value 'Optional' is unsubscriptable (unsubscriptable-object)  (also Union)

# Steps to reproduce

Install Python 3.9, then:

```python
# minimal_repro.py
from typing import Optional, Union
foo: Optional[int] = 3
bar: Union[int, str] = 4
```


### Current behavior

Pylint reports

```
minimal_repro.py:3:5: E1136: Value 'Optional' is unsubscriptable (unsubscriptable-object)
minimal_repro.py:4:5: E1136: Value 'Union' is unsubscriptable (unsubscriptable-object)
```

### Expected behavior

This isn't an error in Python 3.8 so I don't _think_ it should be an error in 3.9?

### pylint --version output

```
pylint 2.6.0
astroid 2.4.2
```
(I tried `pip install pylint astroid --pre -U`)

Tried https://github.com/PyCQA/pylint/pull/3890 on @leycec list:

<summary>
test.py
<details>

```python
import re
import collections
import contextlib

print(tuple[0])
print(list[0])
print(dict[0])
print(set[0])
print(frozenset[0])
print(type[0])
print(collections.deque[0])
print(collections.defaultdict[0])
print(collections.OrderedDict[0])
print(collections.Counter[0])
print(collections.ChainMap[0])
print(collections.abc.Awaitable[0])
print(collections.abc.Coroutine[0])
print(collections.abc.AsyncIterable[0])
print(collections.abc.AsyncIterator[0])
print(collections.abc.AsyncGenerator[0])
print(collections.abc.Iterable[0])
print(collections.abc.Iterator[0])
print(collections.abc.Generator[0])
print(collections.abc.Reversible[0])
print(collections.abc.Container[0])
print(collections.abc.Collection[0])
print(collections.abc.Callable[0])
print(collections.abc.Set[0])
print(collections.abc.MutableSet[0])
print(collections.abc.Mapping[0])
print(collections.abc.MutableMapping[0])
print(collections.abc.Sequence[0])
print(collections.abc.MutableSequence[0])
print(collections.abc.ByteString[0])
print(collections.abc.MappingView[0])
print(collections.abc.KeysView[0])
print(collections.abc.ItemsView[0])
print(collections.abc.ValuesView[0])
print(contextlib.AbstractContextManager[0])
print(contextlib.AbstractAsyncContextManager[0])
print(re.Pattern[0])
print(re.Match[0])
```

</details>
<summary>

It only gives two false positives to be inspected:
```
test.py:10:6: E1136: Value 'type' is unsubscriptable (unsubscriptable-object)
test.py:11:6: E1136: Value 'collections.deque' is unsubscriptable (unsubscriptable-object)
```

--- a/CONTRIBUTORS.txt
+++ b/CONTRIBUTORS.txt
@@ -427,3 +427,5 @@ contributors:
 * Takashi Hirashima: contributor
 
 * Joffrey Mander: contributor
+
+* Julien Palard: contributor

--- a/ChangeLog
+++ b/ChangeLog
@@ -2,6 +2,10 @@
 Pylint's ChangeLog
 ------------------
 
+* Handle class decorators applied to function.
+
+  Closes #3882
+
 * Add check for empty comments
 
 * Fix minor documentation issue in contribute.rst

--- a/doc/development_guide/contribute.rst
+++ b/doc/development_guide/contribute.rst
@@ -110,7 +110,10 @@ your patch gets accepted.
   (`What's New` section). For the release document we usually write some more details,
   and it is also a good place to offer examples on how the new change is supposed to work.
 
-- Add yourself to the `CONTRIBUTORS` file, if you are not already there.
+- Add a short entry in :file:`doc/whatsnew/VERSION.rst`.
+
+- Add yourself to the `CONTRIBUTORS` file, flag youself appropriately
+  (if in doubt, you're a ``contributor``).
 
 - Write a comprehensive commit message
 

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -1724,7 +1724,7 @@ accessed. Python regular expressions are accepted.",
         if inferred is None or inferred is astroid.Uninferable:
             return
 
-        if inferred.decorators:
+        if getattr(inferred, "decorators", None):
             first_decorator = helpers.safe_infer(inferred.decorators.nodes[0])
             if isinstance(first_decorator, astroid.ClassDef):
                 inferred = first_decorator.instantiate_class()

--- a/pylint/checkers/typecheck.py
+++ b/pylint/checkers/typecheck.py
@@ -65,7 +65,7 @@ import astroid
 import astroid.arguments
 import astroid.context
 import astroid.nodes
-from astroid import bases, decorators, exceptions, modutils, objects
+from astroid import bases, decorators, exceptions, helpers, modutils, objects
 from astroid.interpreter import dunder_lookup
 
 from pylint.checkers import BaseChecker, utils
@@ -1720,8 +1720,17 @@ accessed. Python regular expressions are accepted.",
             return
 
         inferred = safe_infer(node.value)
+
         if inferred is None or inferred is astroid.Uninferable:
             return
+
+        if inferred.decorators:
+            first_decorator = helpers.safe_infer(inferred.decorators.nodes[0])
+            if isinstance(first_decorator, astroid.ClassDef):
+                inferred = first_decorator.instantiate_class()
+            else:
+                return  # It would be better to handle function
+                # decorators, but let's start slow.
 
         if not supported_protocol(inferred):
             self.add_message(msg, args=node.value.as_string(), node=node.value)

--- a/tests/checkers/unittest_typecheck.py
+++ b/tests/checkers/unittest_typecheck.py
@@ -300,7 +300,8 @@ class TestTypeChecker(CheckerTestCase):
             def __getitem__(self, item):
                 return item
         """
-        self.typing_option_object_is_subscriptable(decorators)
+        for generic in "Optional", "List", "ClassVar", "Final", "Literal":
+            self.typing_objects_are_subscriptable(generic)
 
         self.decorated_by_a_subscriptable_class(decorators)
         self.decorated_by_an_unsubscriptable_class(decorators)
@@ -308,12 +309,14 @@ class TestTypeChecker(CheckerTestCase):
         self.decorated_by_subscriptable_then_unsubscriptable_class(decorators)
         self.decorated_by_unsubscriptable_then_subscriptable_class(decorators)
 
-    def typing_option_object_is_subscriptable(self, decorators):
+    def typing_objects_are_subscriptable(self, generic):
         module = astroid.parse(
             """
         import typing
-        test = typing.Optional[int]
-        """
+        test = typing.{}[int]
+        """.format(
+                generic
+            )
         )
         subscript = module.body[-1].value
         with self.assertNoMessages():

--- a/tests/checkers/unittest_typecheck.py
+++ b/tests/checkers/unittest_typecheck.py
@@ -287,6 +287,82 @@ class TestTypeChecker(CheckerTestCase):
         with self.assertNoMessages():
             self.checker.visit_subscript(subscript)
 
+    def test_staticmethod_multiprocessing_call(self):
+        """Make sure not-callable isn't raised for descriptors
+
+        astroid can't process descriptors correctly so
+        pylint needs to ignore not-callable for them
+        right now
+
+        Test for https://github.com/PyCQA/pylint/issues/1699
+        """
+        call = astroid.extract_node(
+            """
+        import multiprocessing
+        multiprocessing.current_process() #@
+        """
+        )
+        with self.assertNoMessages():
+            self.checker.visit_call(call)
+
+    def test_not_callable_uninferable_property(self):
+        """Make sure not-callable isn't raised for uninferable
+        properties
+        """
+        call = astroid.extract_node(
+            """
+        class A:
+            @property
+            def call(self):
+                return undefined
+
+        a = A()
+        a.call() #@
+        """
+        )
+        with self.assertNoMessages():
+            self.checker.visit_call(call)
+
+    def test_descriptor_call(self):
+        call = astroid.extract_node(
+            """
+        def func():
+            pass
+
+        class ADescriptor:
+            def __get__(self, instance, owner):
+                return func
+
+        class AggregateCls:
+            a = ADescriptor()
+
+        AggregateCls().a() #@
+        """
+        )
+        with self.assertNoMessages():
+            self.checker.visit_call(call)
+
+    def test_unknown_parent(self):
+        """Make sure the callable check does not crash when a node's parent
+        cannot be determined.
+        """
+        call = astroid.extract_node(
+            """
+        def get_num(n):
+            return 2 * n
+        get_num(10)()
+        """
+        )
+        with self.assertAddsMessages(
+            Message("not-callable", node=call, args="get_num(10)")
+        ):
+            self.checker.visit_call(call)
+
+
+class TestTypeCheckerOnDecorators(CheckerTestCase):
+    "Tests for pylint.checkers.typecheck on decorated functions."
+    CHECKER_CLASS = typecheck.TypeChecker
+
     def test_issue3882_class_decorators(self):
         decorators = """
         class Unsubscriptable:
@@ -417,74 +493,3 @@ class TestTypeChecker(CheckerTestCase):
             )
         ):
             self.checker.visit_subscript(subscript)
-
-    def test_staticmethod_multiprocessing_call(self):
-        """Make sure not-callable isn't raised for descriptors
-
-        astroid can't process descriptors correctly so
-        pylint needs to ignore not-callable for them
-        right now
-
-        Test for https://github.com/PyCQA/pylint/issues/1699
-        """
-        call = astroid.extract_node(
-            """
-        import multiprocessing
-        multiprocessing.current_process() #@
-        """
-        )
-        with self.assertNoMessages():
-            self.checker.visit_call(call)
-
-    def test_not_callable_uninferable_property(self):
-        """Make sure not-callable isn't raised for uninferable
-        properties
-        """
-        call = astroid.extract_node(
-            """
-        class A:
-            @property
-            def call(self):
-                return undefined
-
-        a = A()
-        a.call() #@
-        """
-        )
-        with self.assertNoMessages():
-            self.checker.visit_call(call)
-
-    def test_descriptor_call(self):
-        call = astroid.extract_node(
-            """
-        def func():
-            pass
-
-        class ADescriptor:
-            def __get__(self, instance, owner):
-                return func
-
-        class AggregateCls:
-            a = ADescriptor()
-
-        AggregateCls().a() #@
-        """
-        )
-        with self.assertNoMessages():
-            self.checker.visit_call(call)
-
-    def test_unknown_parent(self):
-        """Make sure the callable check does not crash when a node's parent
-        cannot be determined.
-        """
-        call = astroid.extract_node(
-            """
-        def get_num(n):
-            return 2 * n
-        get_num(10)()
-        """
-        )
-        with self.assertAddsMessages(
-            Message("not-callable", node=call, args="get_num(10)")
-        ):
-            self.checker.visit_call(call)

--- a/tests/checkers/unittest_typecheck.py
+++ b/tests/checkers/unittest_typecheck.py
@@ -318,6 +318,66 @@ class TestTypeChecker(CheckerTestCase):
         with self.assertNoMessages():
             self.checker.visit_subscript(subscript)
 
+    def test_decorated_by_subscriptable_then_unsubscriptable_class_issue3882(self):
+        module = astroid.parse(
+            """
+        class Unsubscriptable:
+            def __init__(self, f):
+                self.f = f
+
+        class Subscriptable:
+            def __init__(self, f):
+                self.f = f
+
+            def __getitem__(self, item):
+                return item
+
+        @Unsubscriptable
+        @Subscriptable
+        def decorated():
+            ...
+
+        test = decorated[None]
+        # TypeError: 'Unsubscriptable' object is not subscriptable
+        """
+        )
+        subscript = module.body[-1].value
+        with self.assertAddsMessages(
+            Message(
+                "unsubscriptable-object",
+                node=subscript.value,
+                args="decorated",
+                confidence=UNDEFINED,
+            )
+        ):
+            self.checker.visit_subscript(subscript)
+
+    def test_decorated_by_unsubscriptable_then_subscriptable_class_issue3882(self):
+        module = astroid.parse(
+            """
+        class Unsubscriptable:
+            def __init__(self, f):
+                self.f = f
+
+        class Subscriptable:
+            def __init__(self, f):
+                self.f = f
+
+            def __getitem__(self, item):
+                return item
+
+        @Subscriptable
+        @Unsubscriptable
+        def decorated():
+            ...
+
+        test = decorated[None]
+        """
+        )
+        subscript = module.body[-1].value
+        with self.assertNoMessages():
+            self.checker.visit_subscript(subscript)
+
     def test_decorated_by_an_unsubscriptable_class_issue3882(self):
         module = astroid.parse(
             """

--- a/tests/checkers/unittest_typecheck.py
+++ b/tests/checkers/unittest_typecheck.py
@@ -303,11 +303,31 @@ class TestTypeChecker(CheckerTestCase):
         for generic in "Optional", "List", "ClassVar", "Final", "Literal":
             self.typing_objects_are_subscriptable(generic)
 
+        self.getitem_on_modules()
         self.decorated_by_a_subscriptable_class(decorators)
         self.decorated_by_an_unsubscriptable_class(decorators)
 
         self.decorated_by_subscriptable_then_unsubscriptable_class(decorators)
         self.decorated_by_unsubscriptable_then_subscriptable_class(decorators)
+
+    def getitem_on_modules(self):
+        """Mainly validate the code won't crash if we're not having a function."""
+        module = astroid.parse(
+            """
+        import collections
+        test = collections[int]
+        """
+        )
+        subscript = module.body[-1].value
+        with self.assertAddsMessages(
+            Message(
+                "unsubscriptable-object",
+                node=subscript.value,
+                args="collections",
+                confidence=UNDEFINED,
+            )
+        ):
+            self.checker.visit_subscript(subscript)
 
     def typing_objects_are_subscriptable(self, generic):
         module = astroid.parse(

--- a/tests/checkers/unittest_typecheck.py
+++ b/tests/checkers/unittest_typecheck.py
@@ -287,40 +287,8 @@ class TestTypeChecker(CheckerTestCase):
         with self.assertNoMessages():
             self.checker.visit_subscript(subscript)
 
-    def test_typing_option_object_is_subscriptable_issue3882(self):
-        module = astroid.parse(
-            """
-        import typing
-        test = typing.Optional[int]
-        """
-        )
-        subscript = module.body[-1].value
-        with self.assertNoMessages():
-            self.checker.visit_subscript(subscript)
-
-    def test_decorated_by_a_subscriptable_class_issue3882(self):
-        module = astroid.parse(
-            """
-        class Deco:
-            def __init__(self, f):
-                self.f = f
-
-            def __getitem__(self, item):
-                return item
-        @Deco
-        def decorated():
-            ...
-
-        test = decorated[None]
-        """
-        )
-        subscript = module.body[-1].value
-        with self.assertNoMessages():
-            self.checker.visit_subscript(subscript)
-
-    def test_decorated_by_subscriptable_then_unsubscriptable_class_issue3882(self):
-        module = astroid.parse(
-            """
+    def test_issue3882_class_decorators(self):
+        decorators = """
         class Unsubscriptable:
             def __init__(self, f):
                 self.f = f
@@ -331,14 +299,51 @@ class TestTypeChecker(CheckerTestCase):
 
             def __getitem__(self, item):
                 return item
+        """
+        self.typing_option_object_is_subscriptable(decorators)
 
+        self.decorated_by_a_subscriptable_class(decorators)
+        self.decorated_by_an_unsubscriptable_class(decorators)
+
+        self.decorated_by_subscriptable_then_unsubscriptable_class(decorators)
+        self.decorated_by_unsubscriptable_then_subscriptable_class(decorators)
+
+    def typing_option_object_is_subscriptable(self, decorators):
+        module = astroid.parse(
+            """
+        import typing
+        test = typing.Optional[int]
+        """
+        )
+        subscript = module.body[-1].value
+        with self.assertNoMessages():
+            self.checker.visit_subscript(subscript)
+
+    def decorated_by_a_subscriptable_class(self, decorators):
+        module = astroid.parse(
+            decorators
+            + """
+        @Subscriptable
+        def decorated():
+            ...
+
+        test = decorated[None]
+        """
+        )
+        subscript = module.body[-1].value
+        with self.assertNoMessages():
+            self.checker.visit_subscript(subscript)
+
+    def decorated_by_subscriptable_then_unsubscriptable_class(self, decorators):
+        module = astroid.parse(
+            decorators
+            + """
         @Unsubscriptable
         @Subscriptable
         def decorated():
             ...
 
         test = decorated[None]
-        # TypeError: 'Unsubscriptable' object is not subscriptable
         """
         )
         subscript = module.body[-1].value
@@ -352,20 +357,10 @@ class TestTypeChecker(CheckerTestCase):
         ):
             self.checker.visit_subscript(subscript)
 
-    def test_decorated_by_unsubscriptable_then_subscriptable_class_issue3882(self):
+    def decorated_by_unsubscriptable_then_subscriptable_class(self, decorators):
         module = astroid.parse(
-            """
-        class Unsubscriptable:
-            def __init__(self, f):
-                self.f = f
-
-        class Subscriptable:
-            def __init__(self, f):
-                self.f = f
-
-            def __getitem__(self, item):
-                return item
-
+            decorators
+            + """
         @Subscriptable
         @Unsubscriptable
         def decorated():
@@ -378,14 +373,11 @@ class TestTypeChecker(CheckerTestCase):
         with self.assertNoMessages():
             self.checker.visit_subscript(subscript)
 
-    def test_decorated_by_an_unsubscriptable_class_issue3882(self):
+    def decorated_by_an_unsubscriptable_class(self, decorators):
         module = astroid.parse(
-            """
-        class Deco:
-            def __init__(self, f):
-                self.f = f
-
-        @Deco
+            decorators
+            + """
+        @Unsubscriptable
         def decorated():
             ...
 

--- a/tests/checkers/unittest_typecheck.py
+++ b/tests/checkers/unittest_typecheck.py
@@ -24,6 +24,7 @@ import astroid
 import pytest
 
 from pylint.checkers import typecheck
+from pylint.interfaces import UNDEFINED
 from pylint.testutils import CheckerTestCase, Message, set_config
 
 try:
@@ -284,6 +285,62 @@ class TestTypeChecker(CheckerTestCase):
         )
         subscript = module.body[-1].value
         with self.assertNoMessages():
+            self.checker.visit_subscript(subscript)
+
+    def test_typing_option_object_is_subscriptable_issue3882(self):
+        module = astroid.parse(
+            """
+        import typing
+        test = typing.Optional[int]
+        """
+        )
+        subscript = module.body[-1].value
+        with self.assertNoMessages():
+            self.checker.visit_subscript(subscript)
+
+    def test_decorated_by_a_subscriptable_class_issue3882(self):
+        module = astroid.parse(
+            """
+        class Deco:
+            def __init__(self, f):
+                self.f = f
+
+            def __getitem__(self, item):
+                return item
+        @Deco
+        def decorated():
+            ...
+
+        test = decorated[None]
+        """
+        )
+        subscript = module.body[-1].value
+        with self.assertNoMessages():
+            self.checker.visit_subscript(subscript)
+
+    def test_decorated_by_an_unsubscriptable_class_issue3882(self):
+        module = astroid.parse(
+            """
+        class Deco:
+            def __init__(self, f):
+                self.f = f
+
+        @Deco
+        def decorated():
+            ...
+
+        test = decorated[None]
+        """
+        )
+        subscript = module.body[-1].value
+        with self.assertAddsMessages(
+            Message(
+                "unsubscriptable-object",
+                node=subscript.value,
+                args="decorated",
+                confidence=UNDEFINED,
+            )
+        ):
             self.checker.visit_subscript(subscript)
 
     def test_staticmethod_multiprocessing_call(self):

